docs/proposals/17-ratelimits-for-shoot-dns-entries.md

```diff
@@ -0,0 +1,112 @@
+
+### Goals
+
+- The rate limit is set and applied separatedly for each `DNSProvider`
+- All update requests caused by `DNSEntries` created or updated from the `shoot-dns-service` should only be forwarded to the backend DNS service, if the update requests are within rate limit of the `DNSProvider` the `DNSEntry` is assigned to.
+- Deletion of `DNSEntries` is not rate limited to avoid freezing on shoot cluster deletion.
+- Updates of `DNSEntries` created by Gardener (e.g. `internal`, `external`, and `ingress`) should not be blocked by the rate limits.
+- If creation/updates of `DNSEntries` of the `shoot-dns-service` are pending because of a rate limit, this should also made be visible in the shoot cluster (either by events or status update)
+
+### Non-Goals
+
+- prioritize updates between `DNSEntries` assigned to the same `DNSProvider`
+
+## Proposal
+
+This proposal involves changes in two components:
+
+- [dns-controller-manager](https://github.com/gardener/external-dns-management) for implementation of rate limiting
+
+- [gardener](https://github.com/gardener/gardener) for configuration of rate limits, applying them to the external `DNSProvider` and annotating `DNSEntries` to exclude then from rate limiting
+
+### DNS controller manager
+
+The spec of a `DNSProvider` is extended by an optional `rateLimit` section.
+
+```go
+type DNSProviderSpec struct {
+    ...
+    // rate limit for create/update operations on DNSEntries assigned to this provider
+    // +optional
+    RateLimit *RateLimit `json:"rateLimit,omitempty"`
+}
+
+type RateLimit struct {
+    // RequestsPerDay is create/update request rate per DNS entry given by requests per day
+    RequestsPerDay int `json:"requestsPerDay"`
+    // Burst allows bursts of up to 'burst' to exceed the rate defined by 'RequestsPerDay', while still maintaining a
+    // smoothed rate of 'RequestsPerDay'
+    Burst int `json:"burst"`
+}
+```
+
+Example manifest:
+
+```yaml
+kind: DNSProvider
+...
+spec:
+  ...
+  rateLimit:
+    requestsPerDay: 120
+    burst: 20
+```
+
+Each valid `DNSEntry` is already automatically assigned to an `DNSProvider` by matching the domain name and the realm.
+This assignment now also includes the regime of the rate limit of this `DNSProvider`.
+The creation or update of DNS records is postponed until the rate limit restrictions are fulfilled.
+The entry stays in the state `Pending` or `Updating` until the request is applied on the DNS cloud service.
+Here the state `Updating` is a new, additional state to be implemented to make transparent that spec and backend are not yet synchronized.
```

Comment:
Can you elaborate why we need `Updating`, i.e., `Pending` is not sufficient?

Reply:
The `Pending` state is only used for newly created `DNSEntries`, .i.e the entry is requested but not yet realised as a DNS record. If a `DNSEntry` is updated, there is currently no state signalling that there is difference between specified and active configuration. Typically the synchronisation happens within seconds, but with the rate limit this could take much longer now. The state `Updating` would signal that a DNS record is already realised, but that it may not reflect the latest changes in the spec.